### PR TITLE
feat: add optional retryOptions to chat completion and chat completion stream requests

### PR DESCRIPTION
**Pull request by iwilltry42**

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -86,67 +89,196 @@ func (c *Client) newRequest(ctx context.Context, method, url string, setters ...
...
+	// RetryAboveCode is the status code above which the request should be retried.
+	RetryAboveCode int
+	RetryCodes     []int
+}
+
+func NewDefaultRetryOptions() RetryOptions {
+	return RetryOptions{
+		Retries:        0,       // = one try, no retries
+		RetryAboveCode: 1,       // any - doesn't matter
+		RetryCodes:     []int{}, // none - doesn't matter
```
Comment (thedadams) [inline client.go]: nit:
```suggestion
		RetryCodes:     nil, // none - doesn't matter
```

File: client.go [hunk 2]
```diff
@@ -86,67 +89,196 @@ func (c *Client) newRequest(ctx context.Context, method, url string, setters ...
...
+		if opt.RetryAboveCode > 0 {
+			r.RetryAboveCode = opt.RetryAboveCode
+		}
+		if len(opt.RetryCodes) > 0 {
+			for _, code := range opt.RetryCodes {
+				if !slices.Contains(r.RetryCodes, code) {
+					r.RetryCodes = append(r.RetryCodes, code)
+				}
+			}
+		}
```
Comment (thedadams) [inline client.go]: nit:
```suggestion
		for _, code := range opt.RetryCodes {
			if !slices.Contains(r.RetryCodes, code) {
				r.RetryCodes = append(r.RetryCodes, code)
			}
		}
```

File: client.go [hunk 3]
```diff
@@ -86,67 +89,196 @@ func (c *Client) newRequest(ctx context.Context, method, url string, setters ...
...
+	}
+}
+
+func (r *RetryOptions) canRetry(statusCode int) bool {
+	if r.RetryAboveCode > 0 && statusCode > r.RetryAboveCode {
+		return true
+	}
+	if len(r.RetryCodes) == 0 {
+		return false
+	}
```
Comment (thedadams) [inline client.go]:
```suggestion
```

File: client.go [hunk 4]
```diff
@@ -86,67 +89,196 @@ func (c *Client) newRequest(ctx context.Context, method, url string, setters ...
...
+	var (
+		resp     *http.Response
+		err      error
+		failures []string
+	)
+
+	// Save the original request body
+	var bodyBytes []byte
+	if req.Body != nil {
+		bodyBytes, err = io.ReadAll(req.Body)
```
Comment (thedadams) [inline client.go:156]:
```suggestion
		bodyBytes, err = io.ReadAll(req.Body)
		_ = req.Body.Close()
```

File: client.go [hunk 5]
```diff
@@ -86,67 +89,196 @@ func (c *Client) newRequest(ctx context.Context, method, url string, setters ...
...
+		// exit on non-retriable status codes
+		if !options.canRetry(resp.StatusCode) {
+			failures = append(failures, fmt.Sprintf("exiting due to non-retriable error in try #%d/%d: %v", i+1, options.Retries+1, resp.StatusCode))
+			return fmt.Errorf("request failed on non-retriable error: %v", strings.Join(failures, "; "))
+		}
+
+		// exponential backoff
+		delay := baseDelay * time.Duration(1<<i)
+		jitter := time.Duration(rand.Int63n(int64(baseDelay)))
+		time.Sleep(delay + jitter)
```
Comment (thedadams) [inline client.go]: nit: if you use this, can you get rid of the check at the top of the loop? You'd also have to add an `outer` scope outside the for-loop.
```suggestion
		select {
			case <-ctx.Done():
				failures = append(failures, fmt.Sprintf("exiting due to canceled context after try #%d/%d: %v", i+1, options.Retries+1, req.Context().Err()))
				break outer
			case <-time.Afer(delay + jitter):
		}
```

File: client.go [hunk 6]
```diff
@@ -86,67 +89,196 @@ func (c *Client) newRequest(ctx context.Context, method, url string, setters ...
...
 	}
-	return resp.Body, nil
+	return fmt.Errorf("request failed %d times: %v", options.Retries+1, strings.Join(failures, "; "))
 }
 
-func sendRequestStream[T streamable](client *Client, req *http.Request) (*streamReader[T], error) {
+func sendRequestStream[T streamable](client *Client, req *http.Request, retryOpts ...RetryOptions) (*streamReader[T], error) {
```
Comment (thedadams) [inline client.go:209]: Same comments from above apply here.

**Review by g-linville — APPROVED** [review]

**Review by thedadams — APPROVED** [review]

**Pull request merged**